atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/asts/SingleBucketSweepTaskIntegrationTest.java

```diff
@@ -0,0 +1,834 @@
+                .as("bucket progress is updated up to the correct point in the queue")
+                .hasValue(BucketProgress.createForTimestampProgress(500L - 1));
+        assertThat(shardProgress.getLastSweptTimestamp(sweepStrategyTestContext.shardAndStrategy()))
+                .as("progress is not updated by the single bucket task")
+                .isEqualTo(-1L);
+
+        // The bucket has not been completed yet, because it is still not fully swept.
+        verify(bucketCompletionListener, never())
+                .markBucketCompleteAndRemoveFromScheduling(
+                        sweepStrategyTestContext.bucketFactory().apply(0));
+    }
+
+    // [v        w][          ]
+    //                  ^ Sweep Timestamp
+    //                  ^ Sweep Task
+    @ParameterizedTest
+    @MethodSource("testContexts")
+    public void doesNotSweepValuesInOtherBuckets(SweepStrategyTestContext sweepStrategyTestContext) {
+        writeTwoCells(sweepStrategyTestContext.dataTable());
+        sweepTimestamp.set(END_OF_BUCKET_ZERO + 4500L);
+
+        assertThat(singleBucketSweepTask.runOneIteration(SweepableBucket.of(
+                        sweepStrategyTestContext.bucketFactory().apply(1), BUCKET_ONE_TIMESTAMP_RANGE)))
+                .as("no entries should have been read from the sweep queue, because this task should not be"
+                        + " sweeping bucket zero")
+                .isEqualTo(0);
+
+        assertThat(keyValueService.get(sweepStrategyTestContext.dataTable(), ImmutableMap.of(DEFAULT_CELL, 250L)))
+                .as("sweep should not have run yet on the older version of the default cell")
+                .hasEntrySatisfying(DEFAULT_CELL, value -> {
+                    assertThat(value.getContents()).isEqualTo(DEFAULT_VALUE);
+                    assertThat(value.getTimestamp()).isEqualTo(100L);
+                });
+
+        assertThat(bucketProgressStore.getBucketProgress(
+                        sweepStrategyTestContext.bucketFactory().apply(0)))
+                .as("bucket progress for other buckets is untouched")
+                .isEmpty();
+        assertThat(bucketProgressStore.getBucketProgress(
+                        sweepStrategyTestContext.bucketFactory().apply(1)))
+                .as("bucket progress is updated up to the sweep timestamp (non-inclusively)")
+                .hasValue(BucketProgress.createForTimestampProgress(4500L - 1L));
+        // The bucket has not been completed yet, because it is still not fully swept.
+        verify(bucketCompletionListener, never()).markBucketCompleteAndRemoveFromScheduling(any());
+    }
+
+    // [v        w]
+    //                  ^ Sweep Timestamp
+    //       ^ Sweep Task
+    @ParameterizedTest
+    @MethodSource("testContexts")
+    public void sweepsBucketAndDeletesBucketGuaranteedNotToBeRecreatedByAssigner(
+            SweepStrategyTestContext sweepStrategyTestContext) {
```

Comment:
So, I was playing around with what happens if startTs is in bucket 1, and commitTs is in some other bucket.
Regardless of what the sweepTimestamp was set (either in between or well after the commitTs), I could not get the bucket to complete.

I need to think about this tomorrow.

Reply:
discussed: yeah this is a problem / sweep would get stuck. (The V1 works for the reason you mention.)